### PR TITLE
[Buttons] Add a custom font property

### PR DESCRIPTION
**Pull request by ianegordon**
Add titleFont property.
Property is compatible with UIAppearance proxy.
Property integrates with Dynamic Type's autoAdjusting property.
Unit Tests and Sample added.

### CODE REVIEW

**Comment by ianegordon** [comment]
Without the dance around systemFontLoader the logic is a decent bit simpler.

**Review by jverkoey — APPROVED** [review]

File: components/Buttons/src/MDCButton.m [hunk 1]
```diff
@@ -738,6 +754,43 @@ - (void)updateBorderWidth {
...
+  if (!font) {
+    // TODO(#2709): Have a single source of truth for fonts
+    // Migrate to [UIFont standardFont] when possible
+    font = [MDCTypography buttonFont];
+  }
+
+  if (_mdc_adjustsFontForContentSizeCategory) {
+    // Dynamic type is enabled so apply scaling
+    font = [font mdc_fontSizedForMaterialTextStyle:MDCFontTextStyleButton
+                              scaledForDynamicType:_mdc_adjustsFontForContentSizeCategory];
```
Comment (jverkoey) [inline components/Buttons/src/MDCButton.m]: Pass YES here?

**Comment by ianegordon** [comment]
Since the +1 from Travis the last commit removes a call to standardFont and simplifies a scale invocation.  If you are happy, I will merge.

File: components/Buttons/tests/unit/ButtonsCustomFontTest.m [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+- (void)testCustomFontTitle {
+  // Given
+  MDCButton *button = [[MDCButton alloc] initWithFrame:CGRectZero];
+
+  // When
+  NSString *customFontName = @"Zapfino";
+  UIFont *customFont = [UIFont fontWithName:customFontName size:14.0];
+  XCTAssert(customFont != nil, @"Unable to instantiate font named %@", customFontName);
+
+  [button setTitleFont:customFont forState:UIControlStateNormal];
```
Comment (randallli) [inline components/Buttons/tests/unit/ButtonsCustomFontTest.m:35]: nit: we usually put the // When here, right before we do the thing we are testing.

File: components/Buttons/tests/unit/ButtonsCustomFontTest.m [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  // Set font to nil
```
Comment (randallli) [inline components/Buttons/tests/unit/ButtonsCustomFontTest.m:54]: nit: move // When here.

File: components/Buttons/tests/unit/ButtonsCustomFontTest.m [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  [button setTitleFont:nil forState:UIControlStateNormal];
+
+  UIFont *finalTitleFont = [button titleFontForState:UIControlStateNormal];
```
Comment (randallli) [inline components/Buttons/tests/unit/ButtonsCustomFontTest.m:57]: nit: This should be below the // Then

File: components/Buttons/src/MDCButton.m [hunk 5]
```diff
@@ -738,6 +750,43 @@ - (void)updateBorderWidth {
...
+  return _fonts[@(state)];
+}
+
+- (void)setTitleFont:(nullable UIFont *)font forState:(UIControlState)state {
+  _fonts[@(state)] = font;
+
+  [self updateTitleFont];
+}
+
+- (void)updateTitleFont {
```
Comment (randallli) [inline components/Buttons/src/MDCButton.m:765]: nit: move into private mark

**Review by randallli — APPROVED** [review]

**Comment by ianegordon** [comment]
TODO: Add a custom font for another state to demonstrate both can be active at once.
TODO: Update comments.
I will handle the TODOs above in a second PR.

**Pull request merged**